renderer/backend/StencilShadowStage.cpp

```diff
@@ -0,0 +1,129 @@
+	idDict defines;
+	defines.Set( "MAX_SHADER_PARAMS", idStr::Fmt("%d", drawBatchExecutor->MaxShaderParamsArraySize<ShaderParams>()) );
+	stencilShadowShader = programManager->LoadFromFiles( "shadows_stencil", 
+		"stages/stencil/stencilshadow.vert.glsl",
+		"stages/stencil/stencilshadow.frag.glsl",
+		defines );
+}
+
+void StencilShadowStage::Shutdown() {}
+
+void StencilShadowStage::DrawStencilShadows( viewLight_t *vLight, const drawSurf_t *shadowSurfs ) {
+	if ( !shadowSurfs || !r_shadows.GetInteger() ) {
+		return;
+	}
+	GL_PROFILE( "StencilShadowPass" );
+
+	if ( r_shadowPolygonFactor.GetFloat() || r_shadowPolygonOffset.GetFloat() ) {
+		qglPolygonOffset( r_shadowPolygonFactor.GetFloat(), -r_shadowPolygonOffset.GetFloat() );
+		qglEnable( GL_POLYGON_OFFSET_FILL );
+	}
+	GL_State( GLS_DEPTHMASK | GLS_COLORMASK | GLS_ALPHAMASK | GLS_DEPTHFUNC_LESS );
+	qglStencilFunc( GL_ALWAYS, 1, 255 );
+
+	if ( glConfig.depthBoundsTestAvailable && r_useDepthBoundsTest.GetBool() ) {
+		qglEnable( GL_DEPTH_BOUNDS_TEST_EXT );
+	}
+	GL_Cull( CT_TWO_SIDED );
+	stencilShadowShader->Activate();
+
+	std::vector<const drawSurf_t*> depthFailSurfs;
+	std::vector<const drawSurf_t*> depthPassSurfs;
+	for (const drawSurf_t *surf = shadowSurfs; surf; surf = surf->nextOnLight) {
+		if (!surf->shadowCache.IsValid()) {
+			continue;
+		}
+
+		bool external = r_useExternalShadows.GetInteger() && !(surf->dsFlags & DSF_VIEW_INSIDE_SHADOW);
+		if (external) {
+			depthPassSurfs.push_back( surf );
+		} else {
+			depthFailSurfs.push_back( surf );
+		}
+	}
+
+	// draw depth-fail stencil shadows
+	qglStencilOpSeparate( backEnd.viewDef->isMirror ? GL_FRONT : GL_BACK, GL_KEEP, GL_DECR_WRAP, GL_KEEP );
+	qglStencilOpSeparate( backEnd.viewDef->isMirror ? GL_BACK : GL_FRONT, GL_KEEP, GL_INCR_WRAP, GL_KEEP );
+	DrawSurfs( depthFailSurfs.data(), depthFailSurfs.size() );
+	// draw traditional depth-pass stencil shadows
+	qglStencilOpSeparate( backEnd.viewDef->isMirror ? GL_FRONT : GL_BACK, GL_KEEP, GL_KEEP, GL_INCR_WRAP );
+	qglStencilOpSeparate( backEnd.viewDef->isMirror ? GL_BACK : GL_FRONT, GL_KEEP, GL_KEEP, GL_DECR_WRAP );
+	DrawSurfs( depthPassSurfs.data(), depthPassSurfs.size() );
+
+	// reset state
+	GL_Cull( CT_FRONT_SIDED );
+	if ( r_shadowPolygonFactor.GetFloat() || r_shadowPolygonOffset.GetFloat() ) {
+		qglDisable( GL_POLYGON_OFFSET_FILL );
+	}
+	if ( glConfig.depthBoundsTestAvailable && r_useDepthBoundsTest.GetBool() ) {
+		qglDisable( GL_DEPTH_BOUNDS_TEST_EXT );
+	}
+	qglStencilOp( GL_KEEP, GL_KEEP, GL_KEEP );
+	// FIXME: move to interaction stage
+	if ( !r_softShadowsQuality.GetBool() || backEnd.viewDef->IsLightGem() /*|| r_shadows.GetInteger()==2 && backEnd.vLight->tooBigForShadowMaps*/ )
+		qglStencilFunc( GL_GEQUAL, 128, 255 ); // enable stencil test - the shadow volume path
+	
+}
+
+void StencilShadowStage::DrawSurfs( const drawSurf_t **surfs, size_t count ) {
+	if (count == 0) {
+		return;
+	}
+
+	vertexCache.VertexPosition( surfs[0]->shadowCache, ATTRIB_SHADOW );
```

Comment:
Actually, I already added a specific function to the vertexCache, because using `VertexPosition` like this is a terrible hack. Just forgot to use it here :)